src/libclipper/src/query_processor.cpp

```diff
@@ -72,70 +74,74 @@ boost::future<Response> QueryProcessor::predict(Query query) {
   log_info_formatted(LOGGING_TAG_QUERY_PROCESSOR, "Found {} tasks",
                      tasks.size());
 
-  vector<boost::future<Output>> task_futures =
+  vector<folly::Future<Output>> task_futures =
       task_executor_.schedule_predictions(tasks);
   if (task_futures.empty()) {
     default_explanation = "No connected models found for query";
     log_error_formatted(LOGGING_TAG_QUERY_PROCESSOR,
                         "No connected models found for query with id: {}",
                         query_id);
   }
-  boost::future<void> timer_future =
-      timer_system_.set_timer(query.latency_budget_micros_);
-
-  boost::future<void> all_tasks_completed;
-  auto num_completed = std::make_shared<std::atomic<int>>(0);
-  std::tie(all_tasks_completed, task_futures) =
-      future::when_all(std::move(task_futures), num_completed);
-
-  auto completed_flag = std::make_shared<std::atomic_flag>();
-  // Due to some complexities of initializing the atomic_flag in a shared_ptr,
-  // we explicitly set the value of the flag to false after initialization.
-  completed_flag->clear();
-  boost::future<void> response_ready_future;
 
-  std::tie(response_ready_future, all_tasks_completed, timer_future) =
-      future::when_either(std::move(all_tasks_completed),
-                          std::move(timer_future), completed_flag);
+  size_t num_tasks = task_futures.size();
 
-  boost::promise<Response> response_promise;
-  auto response_future = response_promise.get_future();
+  // TODO(czumar): Find a more readable name for std::vector<folly::Unit> and
+  // typedef accordingly
+  folly::Future<folly::Unit> timer_future =
+      timer_system_.set_timer(query.latency_budget_micros_);
 
-  // NOTE: We capture the num_completed, completed_flag, and default_explanation
-  // variables
-  // so that they outlive the composed_futures.
-  response_ready_future.then([
-    this, query, query_id, response_promise = std::move(response_promise),
-    selection_state, current_policy, task_futures = std::move(task_futures),
-    num_completed, completed_flag, default_explanation
-  ](auto) mutable {
-
-    vector<Output> outputs;
-    vector<VersionedModelId> used_models;
-    bool all_tasks_timed_out = true;
-    for (auto r = task_futures.begin(); r != task_futures.end(); ++r) {
-      if ((*r).has_exception()) {
-        try {
-          boost::rethrow_exception((*r).get_exception_ptr());
-        } catch (std::exception& e) {
+  std::shared_ptr<std::mutex> outputs_mutex = std::make_shared<std::mutex>();
+  std::vector<Output> outputs;
+  outputs.reserve(task_futures.size());
+  std::shared_ptr<std::vector<Output>> outputs_ptr =
+      std::make_shared<std::vector<Output>>(std::move(outputs));
+
+  std::vector<folly::Future<folly::Unit>> wrapped_task_futures;
```

Comment:
I'm wrapping for error handling as well as aggregating future outputs in the vector owned by the shared pointer `outputs_ptr`. This aggregation is necessary because, in the case that the timer future completes before all of the task futures have completed, we need to access to the values of the completed tasks. `folly::collectAny` doesn't support this behavior directly, hence the wrapping.